README.md
## 网易 pomelo

修改内容
- es6 语法 
- 依赖库升级
    - socket.io 2.0.3


## 使用说明

- pomelo@1.x => pomelo-upgrade@1.x
- pomelo@2.x => pomelo-upgrade@2.x

## 其他插件

- 全局Channel: [pomeloGlobalChannel](https://github.com/frank198/pomeloGlobalChannel)
- flatBuffer支持： [pomelo-flatbuffer](https://github.com/frank198/pomelo-flatbuffer)
- server scale: [pomeloScale](https://github.com/frank198/pomeloScale)


## Pomelo -- a fast, scalable game server framework for node.js

Pomelo is a fast, scalable game server framework for [node.js](http://nodejs.org).
It provides the basic development framework and many related components, including libraries and tools.
Pomelo is also suitable for real-time web applications; its distributed architecture makes pomelo scale better than other real-time web frameworks.

[![Build Status](https://travis-ci.org/NetEase/pomelo.svg?branch=master)](https://travis-ci.org/NetEase/pomelo)

 * Homepage: <http://pomelo.netease.com/>
 * Mailing list: <https://groups.google.com/group/pomelo>
 * Documentation: <http://github.com/NetEase/pomelo>
 * Wiki: <https://github.com/NetEase/pomelo/wiki/>
 * Issues: <https://github.com/NetEase/pomelo/issues/>
 * Tags: game, nodejs

## License

(The MIT License)

Copyright (c) 2012-2017 NetEase, Inc. and other contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



History.md
2.3.1 / 2017-9-18
=================
  * [NEW] 支持Class
  * [NEW] 升级依赖库

1.5.1 / 2017-9-18
=================
  * [NEW] 添加Cron updateCron 命令
  * [NEW] 升级依赖库 ws mqtt

1.4.11 / 2017-4-26
=================
  * [#2 运行报错(pomelo 命令错误)](https://github.com/frank198/pomelo-upgrade/issues/2)

1.4.9 / 2017-4-26
=================
  * [NEW] use es6 update var => let,const

1.4.8 / 2017-4-26
=================
  * [FIX] fix sioconnector bug

1.2.1 / 2015-12-31
================
  * [NEW] upgrade ws to 0.8.0
  * [#771](https://github.com/NetEase/pomelo/pull/771)
  * [#774](https://github.com/NetEase/pomelo/pull/774)
  * [FIX] tls: destory connection when clientError

1.2.0 / 2015-09-18
=================
  * [NEW] upgrade pomelo-admin to 0.4.5
  * [#751](https://github.com/NetEase/pomelo/pull/751)
  * [#741](https://github.com/NetEase/pomelo/pull/741)
  * [#740](https://github.com/NetEase/pomelo/pull/740)
  * [FIX] fix wrong variable in protobuf

1.1.9 / 2015-06-05
=================
  * [NEW] upgrade pomelo-rpc to 0.4.10
  * [NEW] upgrade pomelo-admin to 0.4.4
  * [NEW] upgrade pomelo-logger to 0.1.7

1.1.8 / 2015-05-29
=================
  * fix bug on verison calculating when update proto files
  * avoid modules loading error on windows
  * [NEW] upgrade pomelo-protocol to 0.1.6

1.1.7 / 2015-05-12
=================
  * [#706](https://github.com/NetEase/pomelo/pull/706)
  * [#707](https://github.com/NetEase/pomelo/pull/707)
  * [#443](https://github.com/NetEase/pomelo/pull/443)
  * [#444](https://github.com/NetEase/pomelo/pull/444)
  * [#713](https://github.com/NetEase/pomelo/pull/713)
  * [NEW] upgrade pomelo-rpc to 0.4.9
  * [NEW] upgrade pomelo-admin to 0.4.3

1.1.6 / 2015-03-12
=================
  * [NEW] add configure file automatically reload feature
  * [NEW] add mqtt connector heartbeat timeout option
  * [NEW] upgrade pomelo-rpc to 0.4.8

1.1.5 / 2015-02-26
=================
  * [NEW] upgrade pomelo-rpc to 0.4.7

1.1.4 / 2015-01-23
=================
  [#670](https://github.com/NetEase/pomelo/pull/670)
  [#669](https://github.com/NetEase/pomelo/pull/669)
  [#666](https://github.com/NetEase/pomelo/pull/666)
  [#665](https://github.com/NetEase/pomelo/pull/665)
  [#662](https://github.com/NetEase/pomelo/pull/662)
  [#659](https://github.com/NetEase/pomelo/pull/659)
  [#657](https://github.com/NetEase/pomelo/pull/657)
  [#653](https://github.com/NetEase/pomelo/pull/653)

1.1.2 / 2014-11-12
=================
* [NEW] introduce updateUserInfo for connectionService
  [#637](https://github.com/NetEase/pomelo/pull/637)
* [FIX] fix wrong variable err using 
  [#642](https://github.com/NetEase/pomelo/pull/642)
* [NEW] introduce cancelShutdownHook
  [#644](https://github.com/NetEase/pomelo/pull/644)
* [FIX] revert PR #613, which should not be accepted
  [#649](https://github.com/NetEase/pomelo/pull/649)

1.1.1 / 2014-10-10
=================
* [NEW] upgrade pomelo-protocol to 0.1.4
  [#616](https://github.com/NetEase/pomelo/pull/616)
* [FIX] incorrect this scope
  [#622](https://github.com/NetEase/pomelo/pull/622)
* [FIX] fix bug on arg parse
  [#623](https://github.com/NetEase/pomelo/pull/623)
* [FIX] connection without communication bug

1.1.0 / 2014-09-12
=================
* [NEW] fit for libpomelo2
* [NEW] upgrade pomelo-rpc to 0.4.5
  [#612](https://github.com/NetEase/pomelo/pull/612)
* [FIX] close http server after WebSocketServer.close
  [#613](https://github.com/NetEase/pomelo/pull/613)
* [FIX] update timeout.js
  [#614](https://github.com/NetEase/pomelo/pull/614)
* [FIX] fix typo

1.0.4 / 2014-08-26
=================
* [NEW] upgrade pomelo-rpc to 0.4.3
* [NEW] upgrade pomelo-logger to 0.1.6
* [FIX] pomelo-masterha-plugin reconnect bug miss parameter env
* [#582](https://github.com/NetEase/pomelo/pull/582)

1.0.3 / 2014-07-18
=================
* [NEW] dictVersion : Similar to `protoVersion`, add `dictVersion` and skip sending dict when handshaking if possible.
  [#572](https://github.com/NetEase/pomelo/pull/572)
* [FIX] CRON : Upgrade pomelo-scheduler to v0.3.9. Fix a bug that will loss tasks if The number of days next month is greater than this month.
  [#560](https://github.com/NetEase/pomelo/pull/560)
  [pomelo-scheduler#4](https://github.com/NetEase/pomelo-scheduler/pull/4)
* [FIX] hot update : Remove the error logging if a server doesn't have any handler.
  [#562](https://github.com/NetEase/pomelo/pull/562)
* [NEW] protobuf : Add `protobuf cache` into libpomelo, and add `useProto` handshake option to tell the client whether to use the protobuf.
  [#564](https://github.com/NetEase/pomelo/pull/564)
  [libpomelo#58](https://github.com/NetEase/libpomelo/pull/58)
  [discuss(chinese)](http://nodejs.netease.com/topic/53c6c126898634292c8157a2)

1.0.2 / 2014-07-10
=================
* fix server reconnect bug

1.0.1 / 2014-07-03
=================
* merge pull request #538 #541 #545 #546 #547
* update master watchdog notify method
* upgrade pomelo-rpc to 0.4.2

1.0.0 / 2014-06-19
=================
* mqtt connector
* support ie6,7,8 with sioconnector
* support hot update partially

1.0.0-pre / 2014-05-16
=================
* add udpconnector
* pomelo-rpc load balancing and fault tolerance
* connector wss & tls support
* pomelo-zookeeper-plugin
* pomelo-scale-plugin
* environment directory configuration support
* pomelo-cli dynamic script

0.9.10 / 2014-05-15
=================
* merge pull request #505
* merge pull request #506
* merge pull request from kaisatec
* add getClientAddressBySessionId in sessionService

0.9.9 / 2014-05-06
=================
* merge pull request #495
* merge pull request #499
* merge pull request #501
* upgrade pomelo-admin to 0.3.4

0.9.8 / 2014-05-04
=================
* remove unused module
* update constants definition

0.9.7 / 2014-04-25
=================
* merge pull request #486
* add channelservice rpc error info

0.9.6 / 2014-04-16
=================
* merge pull request #472
* merge pull request #475
* merge pull request #485
* update ssh config

0.9.5 / 2014-04-02

lib/util/Utils.js
/**
 * Created by frank on 16-12-26.
 */
const _ = require('lodash'),
	os = require('os'),
	Constants = require('./constants'),
	util = require('util'),
	logger = require('pomelo-logger').getLogger('pomelo', __filename),
	exec = require('child_process').exec,
	pomelo = require('../pomelo');

class Utils
{
    /**
     *  Invoke callback with check
     * @param {Function} callBack
     * @param args
     */
	static invokeCallback(callBack, ...args)
    {
		if (_.isFunction(callBack))
        {
            // callBack.apply(null, Array.prototype.slice.call(arguments, 1));
			const arg = Array.from ? Array.from(args) : [].slice.call(args);
            // action.apply(null, args);
			callBack(...arg);
		}
	}

    /**
     *  Get the count of elements of object
     * @param obj
     */
	static size(obj)
    {
		let count = 0;
		_.forEach(obj, (objData, index) =>
        {
			if (obj.hasOwnProperty(index) && !_.isFunction(objData))
            {
				count ++;
			}
		});
		return count;
	}

    /**
     * Check a string whether ends with another string
     */
	static endsWith(str, suffix)
    {
		if (!_.isString(str) || !_.isString(suffix) || suffix.length > str.length)
        {
			return false;
		}
		return str.indexOf(suffix, str.length - suffix.length) !== -1;
	}

	static startsWith(str, prefix)
    {
		if (!_.isString(str) || !_.isString(prefix) || prefix.length > str.length)
        {
			return false;
		}
		return str.indexOf(prefix) === 0;
	}

    /**
     * Compare the two arrays and return the difference.
     * @param array1
     * @param array2
     * @returns {Array}
     */
	static arrayDiff(array1, array2)
    {
        // return _.difference(array1, array2);
		const o = {};
		const result = [];
		_.forEach(array2, value =>
        {
			o[value] = true;
		});

		_.forEach(array1, value =>
        {
			if (!o[value])
            {
				result.push(value);
			}
		});
		return result;
	}

	/*
	 * Date format
	 */
	static format(date, format)
    {
		format = format || 'MMddhhmm';
		const month = 3;
		const o = {
			'M+' : date.getMonth() + 1, // month
			'd+' : date.getDate(), // day
			'h+' : date.getHours(), // hour
			'm+' : date.getMinutes(), // minute
			's+' : date.getSeconds(), // second
			'q+' : Math.floor((date.getMonth() + month) / month), // quarter
			'S'  : date.getMilliseconds() // millisecond
		};

		if (/(y+)/.test(format))
        {
			format = format.replace(RegExp.$1, (`${date.getFullYear()}`).substr(4 - RegExp.$1.length));
		}

		_.forEach(o, (time, sign) =>
        {
			if (new RegExp(`(${sign})`).test(format))
            {
				format = format.replace(RegExp.$1, RegExp.$1.length === 1 ? time :
                    (`00${time}`).substr((`${time}`).length));
			}
		});

		return format;
	}

    /**
     * check if has Chinese characters.
     */
	static hasChineseChar(str)
    {
		if (/.*[\u4e00-\u9fa5]+.*$/.test(str))
        {
			return true;
		}
		return false;
	}

    /**
     * transform unicode to utf8
     */
	static unicodeToUtf8(str)
    {
		let i, ch;
		let utf8Str = '';
		const len = str.length;

		for (i = 0; i < len; i++)
        {
			ch = str.charCodeAt(i);

			if ((ch >= 0x0) && (ch <= 0x7F))
            {
				utf8Str += str.charAt(i);

			}
			else if ((ch >= 0x80) && (ch <= 0x7FF))
            {
				utf8Str += String.fromCharCode(0xc0 | ((ch >> 6) & 0x1F));
				utf8Str += String.fromCharCode(0x80 | (ch & 0x3F));

			}
			else if ((ch >= 0x800) && (ch <= 0xFFFF))
            {
				utf8Str += String.fromCharCode(0xe0 | ((ch >> 12) & 0xF));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 6) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | (ch & 0x3F));

			}
			else if ((ch >= 0x10000) && (ch <= 0x1FFFFF))
            {
				utf8Str += String.fromCharCode(0xF0 | ((ch >> 18) & 0x7));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 12) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 6) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | (ch & 0x3F));

			}
			else if ((ch >= 0x200000) && (ch <= 0x3FFFFFF))
            {
				utf8Str += String.fromCharCode(0xF8 | ((ch >> 24) & 0x3));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 18) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 12) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 6) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | (ch & 0x3F));

			}
			else if ((ch >= 0x4000000) && (ch <= 0x7FFFFFFF))
            {
				utf8Str += String.fromCharCode(0xFC | ((ch >> 30) & 0x1));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 24) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 18) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 12) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | ((ch >> 6) & 0x3F));
				utf8Str += String.fromCharCode(0x80 | (ch & 0x3F));
			}
		}
		return utf8Str;
	}

    /**
     *  Ping server to check if network is available
     * @param {string} host
     * @param cb
     */
	static ping(host, cb)
    {
		if (!Utils.isLocal(host))
        {
			const cmd = `ping -w 15 ${host}`;
			exec(cmd, (err, stdout, stderr) =>
            {
				if (err)
                {
					cb(false);
					return;
				}
				cb(true);
			});
		}
		else
        {
			cb(true);
		}
	}

    /**
     * Check if server is exsit.
     *
     */
	static checkPort(server, cb)
    {
		if (!server.port && !server.clientPort)
        {
			Utils.invokeCallback(cb, 'leisure');
			return;
		}
		let port = server.port || server.clientPort;
		const host = server.host;
		const generateCommand = (self, host, port) =>
        {
			let cmd;
			let sshParams = pomelo.app.get(Constants.RESERVED.SSH_CONFIG_PARAMS);
			if (sshParams && Array.isArray(sshParams))
            {
				sshParams = sshParams.join(' ');
			}
			else
            {
				sshParams = '';
			}
			if (!Utils.isLocal(host))
            {
				cmd = util.format('ssh %s %s "netstat -an|awk \'{print $4}\'|grep %s|wc -l"', host, sshParams, port);
			}
			else
            {
				cmd = util.format('netstat -an|awk \'{print $4}\'|grep %s|wc -l', port);
			}
			return cmd;
		};
		const cmd1 = generateCommand(Utils, host, port);
		const child = exec(cmd1, (err, stdout, stderr) =>
        {
			if (err)
            {
				logger.error('command %s execute with error: %j', cmd1, err.stack);
				Utils.invokeCallback(cb, 'error');
			}
			else if (stdout.trim() !== '0')
            {
				Utils.invokeCallback(cb, 'busy');
			}
			else
            {
				port = server.clientPort;
				const cmd2 = generateCommand(Utils, host, port);
				exec(cmd2, function(err, stdout, stderr)
                {
					if (err)
                    {
						logger.error('command %s execute with error: %j', cmd2, err.stack);
						Utils.invokeCallback(cb, 'error');
					}
					else if (stdout.trim() !== '0')
                    {
						Utils.invokeCallback(cb, 'busy');
					}
					else
                    {
						Utils.invokeCallback(cb, 'leisure');
					}
				});
			}
		});
	}

	static isLocal(host)
    {
		const app = require('../pomelo').app;
		if (!app)
        {
			return host === '127.0.0.1' || host === 'localhost' || host === '0.0.0.0' || Utility.InLocal(host);
		}
		return host === '127.0.0.1' || host === 'localhost' || host === '0.0.0.0' || Utility.InLocal(host) || host === app.master.host;
	}

    /**
     * Load cluster server.
     *
     */
	static loadCluster(app, server, serverMap)
    {
		const increaseFields = {};
		const count = parseInt(server[Constants.RESERVED.CLUSTER_COUNT]);
		let seq = app.clusterSeq[server.serverType];
		if (!seq)
        {
			seq = 0;
			app.clusterSeq[server.serverType] = count;
		}
		else
        {
			app.clusterSeq[server.serverType] = seq + count;
		}

		_.forEach(server, (value, key) =>
        {
			if (_.includes(value, Constants.RESERVED.CLUSTER_SIGNAL))
            {
				const base = value.slice(0, -2);
				increaseFields[key] = base;
			}
		});

		for (let i = 0, l = seq; i < count; i++, l++)
        {
			const cserver = _.clone(server); // old clone(server)
			cserver.id = `${Constants.RESERVED.CLUSTER_PREFIX + server.serverType}-${l}`;
			for (const k in increaseFields)
            {
				const v = parseInt(increaseFields[k]);
				cserver[k] = v + i;
			}
			serverMap[cserver.id] = cserver;
		}
	}

	static extends(origin, add)
    {
		if (!add || !_.isObject(add)) return origin;

		const keys = Object.keys(add);
		let i = keys.length;
		while (i--)
        {
			origin[keys[i]] = add[keys[i]];
		}
		return origin;
	}

	static headHandler(headBuffer)
    {
		let len = 0;
		for (let i = 1; i < 4; i++)
        {
			if (i > 1)
				len <<= 8;
			len += headBuffer.readUInt8(i);
		}
		return len;
	}

	static isObject(arg)
    {
		return _.isObject(arg);
	}
}

class Utility
{
	static InLocal(host)
    {
		const localIps = Utility.LocalIps();
		return _.includes(localIps, host);
	}

	static LocalIps()
    {
		const iFaces = os.networkInterfaces();
		const ips = [];
		_.forEach(iFaces, face =>
        {
			_.forEach(face, details =>
            {
				if (details.family === 'IPv4')
                {
					ips.push(details.address);
				}
			});
		});
		return ips;
	}
}

module.exports = Utils;